pkg/sql/sqlutil/internal_executor.go

```diff
@@ -196,16 +287,44 @@ type InternalRows interface {
 	Types() colinfo.ResultColumns
 }
 
-// SessionBoundInternalExecutorFactory is a function that produces a "session
-// bound" internal executor.
-type SessionBoundInternalExecutorFactory func(
+// InternalExecutorFactory is a function that produces an internal executor.
+type InternalExecutorFactory func(
 	context.Context, *sessiondata.SessionData,
 ) InternalExecutor
 
+// InternalExecutorFactoryWithTxn is a function to construct an internal executor
+// that can run queries within a txn.
+type InternalExecutorFactoryWithTxn func(
+	ctx context.Context,
+	data *sessiondata.SessionData,
+	collection txnbase.DescsCollection,
+	records map[txnbase.DescpbID]txnbase.JobRecords,
+) InternalExecutor
```

Comment:
I don't understand why this is needed. All of the users can either import sql (planner) or have the dependency to construct the internal executor directly injected.

Reply:
If we don't want this and we keep the original `sqlutil.InternalExecutorFactory`, which is a function returning an interface `sqlutil.InternalExecutor`, and if we want to pass these extra txn state info, I think we have to create a `SetExtraTxnState()` func that hangs of `sqlutil.InternalExecutor`, which I think we discussed as not ideal. 

The reason I create this `InternalExecutorFactoryWithTxn` new type is to make sure the txn related info are more tightly coupled with the creation of an internal executor.

> All of the users can either import sql (planner) or have the dependency to construct the internal executor directly injected.

I'm not sure if I understand this correctly, could you say more?
I'm worried that import `sql` be difficult to import since it may cause cylic dependency. At least in `pkg/sql/catalog/descs/txn.go` where we define `CollectionFactory.TxnWithExecutor()` (where we create an internal executor with a descriptor collection created by the collection factory), if we have to import `sql`, it will create the following dependency loop:
```
.-> //pkg/jobs:jobs (6956aae3f4db2488eca6e16211ad41efab883e20a3a6b337a4fbb2a07bc91ca5)
|   //pkg/sql/catalog/descs:descs (6956aae3f4db2488eca6e16211ad41efab883e20a3a6b337a4fbb2a07bc91ca5)
|   //pkg/sql:sql (6956aae3f4db2488eca6e16211ad41efab883e20a3a6b337a4fbb2a07bc91ca5)
|   //pkg/sql/schemachanger/scrun:scrun (6956aae3f4db2488eca6e16211ad41efab883e20a3a6b337a4fbb2a07bc91ca5)
|   //pkg/sql/schemachanger/scexec:scexec (6956aae3f4db2488eca6e16211ad41efab883e20a3a6b337a4fbb2a07bc91ca5)
`-- //pkg/jobs:jobs (6956aae3f4db2488eca6e16211ad41efab883e20a3a6b337a4fbb2a07bc91ca5)
```

Reply:
My recollection of our conversation last week was that we'd want to have a function similar to the `InternalExecutorFactory` function we have today be a dependency of the `descs` package such that `descs.CollectionFactory` can create an `InternalExecutor`. It's definitely the case that `descs` cannot import `sql` directly, however, the `server` package can inject the `sql` package logic into the `descs.CollectionFactory` when it is constructed. In this way, `descs` will not need to import `sql`. The type of the injected function does not need to exist in either `sqlutil` or `sql`, it can be just in the signature of `descs.NewCollectionFactory`

Reply:
Thanks for the explanation! Yes, it makes sense to me that we can have a `descs.InternalExecutorFactory` type, have an `IEFactory` field of the `CollectionFactory` struct, and let it initialized in `server_sql.go`, just as how `distSQLServer.ServerConfig.InternalExecutorFactory` is initialized. 

My last concern is about adapt the current use cases to this new interface. IIUC we want to replace this kind of use case 
```go
func someFunc(db *kv.DB, ieFactory sqlutil.InternalExecutorFactory)  error {
    return db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
        ie := ieFactory(...)
        return ie.ExecEx(...)
    }
  )
}
```

into 

```go
func someFunc(db *kv.DB, colFactory descs.CollectionFactory)  error {
    return colFactory.TxnWithExecutor(ctx, func(ctx context.Context, txn *kv.Txn, ie sqlutil.InternalExecutor) error {
        return ie.ExecEx(...)
    }
  )
}
```
It means that we will likely need to replace fields of type `sqlutil.InternalExecutorFactory` or `sqlutil.InternalExecutor` with `descs.CollectionFactory`. Since objects of these former 2 types are so widely used and often passed freely between functions and packages, I'm not sure if `descs` can always be accessible in these contexts. I think at least we need to prepare ourselves for some ugly/big refactorization.

Reply:
big != ugly, but I'm also not sure that anything we're talking about is that big. I see 6 uses of `SessionBoundInternalExecutorFactory`, which doesn't sound like very many to me. It'd be helpful if you could enumerate the cases of concern.

Reply:
My way is to look at the usages of internal executor's query functions (e.g. `ie.QueryRowEx()`) with not-nil `txn`, and trace upward where the `txn` is from. There're so many (~100) use cases of this pattern, and some of them seem to me not easy to adapt the new interface with collection factory.

For example, [this usage ](https://github.com/cockroachdb/cockroach/blob/d7b901d7d28ceb4900dd60c5c73b2180089a3e3a/pkg/cloud/userfile/filetable/file_table_read_writer.go#L471) of `ie.ExecEx()` in `filetable/file_table_read_writer.go`. Refactoring this use case may require importing `descs` to `cloud` for `ExternalStorageContext`, which will result in a cyclic dependency.

Reply:
@ajwerner I've made changes to how `CollectionFactory.ieFactoryWithTxn` is initialized. However, we still need to create an interface in a separated package for `jobs.Record` for passing schema change job records to the internal executor , since importing `jobs` in `descs` will cause dependency loop. But indeed it looks better now!